karapace/protobuf/serialization.py

```diff
@@ -0,0 +1,316 @@
+        result.append(
+            OptionElement("java_generate_equals_and_hash", OptionElement.Kind.BOOLEAN, options.java_generate_equals_and_hash)
+        )
+    if options.HasField("deprecated"):
+        result.append(OptionElement("deprecated", OptionElement.Kind.BOOLEAN, options.deprecated))
+    if options.HasField("java_string_check_utf8"):
+        result.append(OptionElement("java_string_check_utf8", OptionElement.Kind.BOOLEAN, options.java_string_check_utf8))
+    if options.HasField("cc_enable_arenas"):
+        result.append(OptionElement("cc_enable_arenas", OptionElement.Kind.BOOLEAN, options.cc_enable_arenas))
+    if options.HasField("objc_class_prefix"):
+        result.append(OptionElement("objc_class_prefix", OptionElement.Kind.STRING, options.objc_class_prefix))
+    if options.HasField("csharp_namespace"):
+        result.append(OptionElement("csharp_namespace", OptionElement.Kind.STRING, options.csharp_namespace))
+    if options.HasField("swift_prefix"):
+        result.append(OptionElement("swift_prefix", OptionElement.Kind.STRING, options.swift_prefix))
+    if options.HasField("php_class_prefix"):
+        result.append(OptionElement("php_class_prefix", OptionElement.Kind.STRING, options.php_class_prefix))
+    if options.HasField("php_namespace"):
+        result.append(OptionElement("php_namespace", OptionElement.Kind.STRING, options.php_namespace))
+    if options.HasField("php_generic_services"):
+        result.append(OptionElement("php_generic_services", OptionElement.Kind.BOOLEAN, options.php_generic_services))
+    if options.HasField("php_metadata_namespace"):
+        result.append(OptionElement("php_metadata_namespace", OptionElement.Kind.STRING, options.php_metadata_namespace))
+    if options.HasField("ruby_package"):
+        result.append(OptionElement("ruby_package", OptionElement.Kind.STRING, options.ruby_package))
+    return result
+
+
+def deserialize(schema_b64: str) -> ProtoFileElement:
+    serialized_pb = base64.b64decode(schema_b64, validate=True)
+    proto = google.protobuf.descriptor_pb2.FileDescriptorProto()
+    proto.ParseFromString(serialized_pb)
+    imports: list[TypeName] = []
+    public_imports: list[TypeName] = []
+    for index, dep in enumerate(proto.dependency):
+        if index in proto.public_dependency:
+            public_imports.append(TypeName(dep))
+        else:
+            imports.append(TypeName(dep))
+    types: list[TypeElement] = []
+    for enumtype in proto.enum_type:
+        types.append(_deserialize_enum(enumtype))
+    for msgtype in proto.message_type:
+        types.append(_deserialize_msg(msgtype))
+    options: list[OptionElement] = _deserialize_options(proto.options)
+
+    return ProtoFileElement(
+        DEFAULT_LOCATION,
+        package_name=PackageName(proto.package),
+        syntax=Syntax(proto.syntax),
+        imports=imports,
+        public_imports=public_imports,
+        types=types,
+        options=options,
+    )
+
```

Comment:
The tests are targeting only that function. Since the implementation it's complex and figuring out every detail it's kinda tricky, doesn't make sense to unit test also the nested functions?


Like for e.g. creating unit tests for the `_deserialize_enum` the `_deserialize_msg` and the `_deserialize_options`.

I know that being able to deserialize the whole message imply being able to deserialize the components, but IMO would enable avoid regressions in the sub components and having a more understandable code by looking in the tests

Reply:
I think adding unit tests for those internal functions makes refactoring unnecessary hard, and I would avoid it.  The coverage for sub components is achieved by multiple schemas used with unit tests.

Reply:
Good point, let's just make sure that we are hitting the various branches